snapshots/windows/windows.go

```diff
@@ -202,14 +205,58 @@ func (s *snapshotter) Commit(ctx context.Context, name, key string, opts ...snap
 	// grab the existing id
 	id, _, _, err := storage.GetInfo(ctx, key)
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "failed to get storage info for %s", key)
 	}
 
-	usage, err := fs.DiskUsage(ctx, s.getSnapshotDir(id))
+	snapshot, err := storage.GetSnapshot(ctx, key)
 	if err != nil {
 		return err
 	}
 
+	path := s.getSnapshotDir(id)
+
+	if strings.Contains(key, snapshots.UnpackKeyPrefix) {
+		// Do nothing, ociwclayer.ImportLayerFromTar has left this layer in its final state
+	} else {
+		// Reimport the layer over itself, to transfer the files from the sandbox.vhdx to the on-disk storage.
+
+		// TODO darrenstahlmsft: When this is done isolated, we should disable these.
+		// it currently cannot be disabled, unless we add ref counting. Since this is
+		// temporary, leaving it enabled is OK for now.
+		// https://github.com/containerd/containerd/issues/1681
+		if err = winio.EnableProcessPrivileges([]string{winio.SeBackupPrivilege, winio.SeRestorePrivilege}); err != nil {
+			return errors.Wrap(err, "failed to enable necessary privileges")
+		}
+
+		parentLayerPaths := s.parentIDsToParentPaths(snapshot.ParentIDs)
+		reader, writer := io.Pipe()
+
+		go func() {
+			err := ociwclayer.ExportLayerToTar(ctx, writer, path, parentLayerPaths)
```

Comment:
is it long running process? it might lock other snapshot's actions since botldb only allows one writer tx.

Reply:
I assume the boltdb lock is from `TransactionContext` earlier?

Hmm, interesting. I'm not sure what the alternative options are.

If we try to do these operations outside the lock, we need to be in a new state, where we're not yet committed (as new layers based on this will fail) but we're also not active (as trying to commit this layer will fail, and mounting the layer would let you modify the data during/after commit, assuming it doesn't just fail to mount in HCS).

We already have a risk here that if this operation fails after we start the `ImportLayerFromTar` part, we can't actually roll-back the filesystem without intrusive knowledge of the WCOW layer structure. Releasing the lock during the IO operatoin sounds much worse.

Only allowing one writer _total_ (rather than one per snapshot) is surprising, actually. I see that as far as `Commit` operations go, only `devmapper` does anything interesting in its implementation (flushing any pending IO operations), the rest are only updating metadata.

Reply:
To address the real-life speed effect, the time taken is going to be proportional to the size of the layer. The large base layers, e.g. the 5GB or so Server Core base layers, aren't affected, because they hit the "unpack" code-path earlier.

On the other hand, my "break it all" test case would be https://github.com/adamrehn/ue4-docker/, which produces some intermediate layers in the 100GB+ range. A full build of UE4 4.26 with debug info requires about 600-700GB of free disk space, but part of that is the heavy on-disk duplication inherent in the WCOW filesystem storage format: during this process in particular, there's three copies of the layer's data on-disk, and afterwards, there's two copies.

The commit operation under Docker for such large layers can take a long time (I've never timed it, but if someone said "an hour", I'd believe it) which is probably too-long to hold a metadata lock that blocks all other container-commit operations.

Reply:
Good to know this! I think we can consider other process to improve the performance issue about huge layer. It should be follow-up pr.

Reply:
I opened #5384 for this.